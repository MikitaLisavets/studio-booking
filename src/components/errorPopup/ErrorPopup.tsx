import styles from './ErrorPopup.module.scss';
import React, { useState, useEffect } from 'react';
import { ErrorRequest } from '../../types/types';

interface ErrorPopupProps {
  error: ErrorRequest | null;
  onClose: () => void;
}

export default function ErrorPopup(props: ErrorPopupProps): JSX.Element {
  const { error, onClose } = props;
  const [visible, setVisible] = useState(!!error);

  const visibleStyles = visible ? styles.visible : '';
  const popupStyles = [styles.popup, visibleStyles].join(' ');

  function hidePopup(): void {
    console.log('hhhhh');
    setVisible(false);
    onClose();
  }

  useEffect(() => {
    if (!error) return;
    setVisible(true);
  }, [error]);

  return (
    <div className={popupStyles}>
      <button className={styles.close} onClick={hidePopup}>âœ•</button>
      <p>{error?.message}</p>
    </div>
  );
}