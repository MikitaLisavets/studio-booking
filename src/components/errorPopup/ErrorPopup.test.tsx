import ErrorPopup from './ErrorPopup';
import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

function createComponent(newProps = {}): ShallowWrapper {
  const props = {
    error: null,
    onClose: (): void => undefined
  };

  const updatedProps = { ...props, ...newProps };

  return shallow(
    <ErrorPopup {...updatedProps}/>
  );
}

describe('ErrorPopup', () => {
  let component: ShallowWrapper;

  describe('when there is no error', () => {
    beforeAll(() => {
      component = createComponent();
    });

    it('renders correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('does not have "visible" class', () => {
      expect(component.find('.popup')).not.toHaveClassName('visible');
    });
  });

  describe('when there is error', () => {
    const onCloseMock = jest.fn();
    
    beforeAll(() => {
      component = createComponent({
        error: { message: 'errorMessage' },
        onClose: onCloseMock
      });
    });

    it('renders correctly', () => {
      expect(component).toMatchSnapshot();
    });

    it('adds "visible" class', () => {
      expect(component.find('.popup')).toHaveClassName('visible');
    });

    it('shows error message', () => {
      expect(component.find('.popup p')).toHaveText('errorMessage');
    });

    describe('and then clicking on close button', () => {
      beforeAll(() => {
        component.find('.close').simulate('click');
      });

      it('removes "visible" class', () => {
        expect(component.find('.popup')).not.toHaveClassName('visible');
      });
  

      it('calls onClick handler', () => {
        expect(onCloseMock).toHaveBeenCalled();
      });
    });
  });
});
