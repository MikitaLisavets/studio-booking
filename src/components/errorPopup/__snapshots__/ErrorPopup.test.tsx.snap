// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorPopup when there is error renders correctly 1`] = `
<div
  className="popup visible"
>
  <button
    className="close"
    onClick={[Function]}
  >
    ✕
  </button>
  <p>
    errorMessage
  </p>
</div>
`;

exports[`ErrorPopup when there is no error renders correctly 1`] = `
<div
  className="popup "
>
  <button
    className="close"
    onClick={[Function]}
  >
    ✕
  </button>
  <p />
</div>
`;
